File: src/utils/protocols/industrial.ts
import { Protocol } from '../types';

export const IndustrialProtocols: Protocol[] = [
  {
    name: 'Modbus',
    ports: [502],
    description: 'Modbus TCP',
    scanFunction: async (host: string, port: number) => {
      // Modbus detection logic
      return {
        protocol: 'Modbus',
        version: 'TCP',
        mode: 'TCP/IP'
      };
    }
  },
  {
    name: 'DNP3',
    ports: [20000],
    description: 'Distributed Network Protocol',
    scanFunction: async (host: string, port: number) => {
      // DNP3 detection logic
      return {
        protocol: 'DNP3',
        version: '3.0',
        mode: 'TCP/IP'
      };
    }
  },
  {
    name: 'BACnet',
    ports: [47808],
    description: 'Building Automation and Control Networks',
    scanFunction: async (host: string, port: number) => {
      // BACnet detection logic
      return {
        protocol: 'BACnet',
        version: 'IP',
        mode: 'BACnet/IP'
      };
    }
  }
];

File: src-tauri/src/protocols/industrial.rs
```rust
use std::net::TcpStream;
use std::time::Duration;
use serde::{Serialize, Deserialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct IndustrialProtocolResult {
    pub protocol: String,
    pub version: Option<String>,
    pub details: ProtocolDetails,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ProtocolDetails {
    pub vendor: Option<String>,
    pub model: Option<String>,
    pub firmware: Option<String>,
    pub security_level: Option<String>,
}

pub struct IndustrialProtocols;

impl IndustrialProtocols {
    pub async fn detect_modbus(host: &str, port: u16) -> Result<IndustrialProtocolResult, String> {
        // Modbus detection implementation
        let mut result = IndustrialProtocolResult {
            protocol: "Modbus".to_string(),
            version: Some("TCP".to_string()),
            details: ProtocolDetails {
                vendor: None,
                model: None,
                firmware: None,
                security_level: Some("Basic".to_string()),
            },
        };

        if let Ok(stream) = TcpStream::connect((host, port)) {
            stream.set_read_timeout(Some(Duration::from_secs(2)))?;
            // Implement Modbus protocol detection logic
        }

        Ok(result)
    }

    pub async fn detect_dnp3(host: &str, port: u16) -> Result<IndustrialProtocolResult, String> {
        // DNP3 detection implementation
        Ok(IndustrialProtocolResult {
            protocol: "DNP3".to_string(),
            version: Some("3.0".to_string()),
            details: ProtocolDetails {
                vendor: None,
                model: None,
                firmware: None,
                security_level: Some("Medium".to_string()),
            },
        })
    }

    pub async fn detect_bacnet(host: &str, port: u16) -> Result<IndustrialProtocolResult, String> {
        // BACnet detection implementation
        Ok(IndustrialProtocolResult {
            protocol: "BACnet".to_string(),
            version: Some("IP".to_string()),
            details: ProtocolDetails {
                vendor: None,
                model: None,
                firmware: None,
                security_level: Some("Low".to_string()),
            },
        })
    }
}
```